game/gamestate.go

```diff
@@ -147,3 +539,263 @@ func (gs *GameState) AreAdjacent(cantonID1, cantonID2 int) bool {
+		if gameMove.ActionType == AttackAction {
+			// Apply attack move
+			err := newGs.Attack(gameMove.FromCantonID, gameMove.ToCantonID)
+			if err != nil {
+				// Panic on error
+				panic(err)
+			}
+		} else if gameMove.ActionType == PassAction {
+			// End attack phase
+			newGs.AdvancePhase()
+		} else {
+			// Invalid action for this phase
+			panic("Invalid action for AttackPhase")
+		}
+	case ManeuverPhase:
+		if gameMove.ActionType == ManeuverAction {
+			// Apply maneuver move
+			err := newGs.MoveTroops(gameMove.FromCantonID, gameMove.ToCantonID, gameMove.NumTroops)
+			if err != nil {
+				// Panic on error
+				panic(err)
+			}
+			// After one maneuver, end the phase
+			newGs.AdvancePhase()
+		} else if gameMove.ActionType == PassAction {
+			// End maneuver phase
+			newGs.AdvancePhase()
+		} else {
+			// Invalid action for this phase
+			panic("Invalid action for ManeuverPhase")
+		}
+	default:
+		panic("Unknown game phase")
+	}
+
+	// Update the last move
+	newGs.LastMove = move
+	return newGs
+}
+
+// advancePhase moves the game to the next phase or next player's turn.
+func (gs *GameState) AdvancePhase() {
+	switch gs.Phase {
+	case InitialPlacementPhase:
+		gs.CurrentPlayer = gs.NextPlayer()
+		if gs.PlayerTroops[1] == 0 && gs.PlayerTroops[2] == 0 {
+			gs.Phase = ReinforcementPhase
+			gs.HandleCardTrading()
+			gs.calculateTroopsToPlace()
+		}
+	case ReinforcementPhase:
+		gs.Phase = AttackPhase
+	case AttackPhase:
+		gs.Phase = ManeuverPhase
+	case ManeuverPhase:
+		gs.AwardCardIfEligible()
+		gs.Phase = ReinforcementPhase
+		gs.CurrentPlayer = gs.NextPlayer()
+		gs.HandleCardTrading()
+		gs.calculateTroopsToPlace()
+	}
+}
+
+func (gs *GameState) getEnemyAdjacentTerritories() []int {
+	territoriesSet := make(map[int]struct{})
+	for cantonID, owner := range gs.Ownership {
+		if owner == gs.CurrentPlayer {
+			for _, adjID := range gs.Map.Cantons[cantonID].AdjacentIDs {
+				if gs.Ownership[adjID] != gs.CurrentPlayer {
+					territoriesSet[cantonID] = struct{}{}
+					break
+				}
+			}
+		}
+	}
+	var territories []int
+	for t := range territoriesSet {
+		territories = append(territories, t)
+	}
+	return territories
+}
+
+// calculateTroopsToPlace calculates the number of troops the current player should place.
+func (gs *GameState) calculateTroopsToPlace() {
+	numTerritories := 0
+	troops := 0
+
+	// Count owned territories
+	for _, owner := range gs.Ownership {
+		if owner == gs.CurrentPlayer {
+			numTerritories++
+		}
+	}
+	troops += max(3, numTerritories/3)
+
+	// Check for region bonuses
+	for _, region := range gs.Map.Regions {
+		ownsAll := true
+		for _, cantonID := range region.CantonIDs {
+			if gs.Ownership[cantonID] != gs.CurrentPlayer {
+				ownsAll = false
+				break
+			}
+		}
+		if ownsAll {
+			troops += BONUS_TROOPS_REGION
+		}
+	}
+	gs.TroopsToPlace = troops
+}
+
+func (gs *GameState) NextPlayer() int {
+	if gs.CurrentPlayer == 1 {
+		return 2
+	}
+	return 1
+}
+
+func (gs *GameState) Winner() string {

```

Comment:
- [x] rename to `isTerminal() bool ` (check at the end of `Play()` after state transition)
- [x] expose winner through `Winner() string` (getter for `winnner` field with default empty value)
```suggestion
func (gs GameState) isTerminal() bool {
```